ngraph/test/runtime/reference/ctc_loss.hpp

```diff
@@ -0,0 +1,164 @@
+                            std::to_string(actualTargetLen));
+                    }
+
+                    const U* target = &labels[b * maxTime];
+                    // Decoding target:
+                    // merge repeated characters if preprocess_collapse_repeated == True,
+                    // find unique elemnts if unique == True
+                    size_t decodedTargetLen = 0lu;
+                    if (unique)
+                    {
+                        std::unordered_set<U> uniqVals;
+                        for (size_t t = 0lu; t < actualTargetLen; t++)
+                        {
+                            if (uniqVals.find(target[t]) != uniqVals.end())
+                            {
+                                continue;
+                            }
+                            uniqVals.insert(target[t]);
+                            targetD[decodedTargetLen++] = target[t];
+                        }
+                    }
+                    else if (preprocessCollapseRepeated)
+                    {
+                        U prevValue = target[0];
+                        targetD[decodedTargetLen++] = target[0];
+                        for (size_t t = 1lu; t < actualTargetLen; t++)
+                        {
+                            if (target[t] == prevValue)
+                            {
+                                continue;
+                            }
+                            targetD[decodedTargetLen++] = target[t];
+                            prevValue = target[t];
+                        }
+                    }
+                    else
+                    {
+                        std::copy(target, target + actualTargetLen, targetD.data());
+                        decodedTargetLen = actualTargetLen;
+                    }
+
+                    const size_t BTC = b * TC;
+
+                    std::vector<T> kExp(actualLogitLen, 0);
+                    for (size_t t = 0; t < actualLogitLen; t++)
+                    {
+                        for (size_t c = 0; c < classesNum; c++)
+                        {
+                            kExp[t] += std::exp(logits[BTC + classesNum * t + c]);
+                        }
+                    }
+
+                    T res = 0;
+
+                    // Looking for aligned paths
+                    std::function<void(size_t targetIdx, size_t start, size_t end)> findPaths =
+                        [&](size_t targetIdx, size_t start, size_t end) {
+                            if (end > actualLogitLen)
+                            {
+                                T ex = 0;
+                                T denom = 1;
+                                for (size_t t = 0; t < actualLogitLen; t++)
+                                {
+                                    ex += logits[BTC + classesNum * t + pathS[t]];
+                                    denom *= kExp[t];
```

Comment:
Use log(softmax(logit)) to compute log(prob(seq)) as a sum of log(softmax(...)) of logits. After that please sum up log(prob)) for sequences using formula ln(a + b) = ln(a) + ln(1 + exp(ln(b) - ln(a))): https://github.com/githubharald/CTCWordBeamSearch/issues/23

Reply:
Fixed